caproto/threading/client.py

```diff
@@ -300,6 +309,240 @@ def __call__(self):
                     # sent to disconnected sockets
 
 
+class SearchResults:
+    '''
+    Thread-safe handling of all past and in-process search results
+
+    Acts partially as a container type which is keyed on PV name, such that the
+    following are possible::
+
+        1. `SearchResults()[name] -> [time, address]`
+        2. `name in SearchResults()`
+
+    Attributes
+    ----------
+    name_to_addrs : dict
+        Holds search results.
+        Maps name -> [(address, name, time), ...]
+    addr_to_name : dict
+        Holds search results.
+        Maps address -> {name1, name2, ...}
+    _unanswered_searches : dict
+        Holds pending searches
+        Maps search_id -> [name, results_queue, retirement_deadline]
+    _lock : threading.RLock
+        Lock for internal updates to SearchResults status
+    _search_id_counter : ThreadsafeCounter
+        Counter for new searches. This will be kept in sync with
+        _unanswered_searches such that there is no overlap in keys.
+    '''
+
+    def __init__(self):
+        self._lock = threading.RLock()
+        # map name to (time, address)
+        self.name_to_addrs = defaultdict(dict)
+        self.addr_to_names = defaultdict(set)
+        self._searches = {}
+        self._searches_by_name = {}
+        self._unanswered_searches = {}
+        self._search_id_counter = ThreadsafeCounter(
+            initial_value=random.randint(0, MAX_ID),
+            dont_clash_with=self._unanswered_searches,
+            lock=self._lock,
+        )
+
+    def _locked(func):
+        @functools.wraps(func)
+        def wrapped(self, *args, **kwargs):
+            with self._lock:
+                return func(self, *args, **kwargs)
+        return wrapped
+
+    @_locked
+    def new_server_found(self, address):
+        '''
+        Call when a new server beacon is found:
+        Bring all the unanswered searches out of retirement to see if we have a
+        new match.
+        '''
+        retirement_deadline = time.monotonic() + SEARCH_RETIREMENT_AGE
+        for item in self._unanswered_searches.values():
+            item[-1] = retirement_deadline
+
+    @property
+    @_locked
+    def unanswered_searches(self):
+        'All unanswered searches'
+        return dict(self._unanswered_searches)
```

Comment:
To save from copying on access, it might be worth stealing this `Frozen` class from xarray: https://github.com/pydata/xarray/blob/3fb22cbf86e49be4a444fa062350bf7570abd6da/xarray/core/utils.py#L333-L356

Reply:
I think we can just remove this altogether. This is a remnant of previous, less-clean abstraction that required `Broadcaster` to reach into `SearchResults.unanswered_searches`.